agent/agent.go

```diff
@@ -89,8 +90,17 @@ var (
 	monitoringResourceType   = flag.String("monitoring-resource-type", "gce_instance", "The monitoring resource type. Eg: gce_instance")
 	monitoringResourceLabels = flag.String("monitoring-resource-labels", "", "Comma separated key value pairs specifying the resource labels. Eg: 'instance-id=12345678901234,instance-zone=us-west1-a")
 
+	enableGracefulShutdown  = flag.Bool("enable-graceful-shutdown", false, "Whether or not to enable graceful shutdown")
+	gracefulShutdownTimeout = flag.Duration("graceful-shutdown-timeout", 30*time.Second, "Timeout for graceful shutdown")
+
 	sessionLRU    *sessions.Cache
 	metricHandler *metrics.MetricHandler
+
+	shutdownListenerMu          sync.Mutex
```

Comment:
This logic with the multiple channels, mutex, and wait group is very complicated.

I'd like to suggest a different pattern following the precedent the health checks use.

The health checks run completely separately from the main flow of control, and when they fail enough times they simply short-circuit everything else by calling `log.Fatal`.

I think we could do something like that here:

1. Pass a context into the `pollForNewRequests` method that is used to terminate the polling loop. When that context's `Done` channel is closed, the loop returns (however, this context would *not* be used for processing the individual requests returned from a successful poll).
2. Construct this context in the helper method that intercepts the termination signals. Once one of those signals is received (inside a separate goroutine), cancel the context and then start a timer. Once that timer runs out call `log.Fatal` to kill off any remaining request processing.

I think this approach would give us the same behavior, but with much fewer things to keep track of.

Reply:
> I think this approach would give us the same behavior, but with much fewer things to keep track of.

Thanks for the review and new approach suggestion Omar. While there will surely be lesser moving parts with this approach, the behaviour would not be same. 
In current design of the PR changes, we do an optimized waiting during the graceful shutdown. We do not always wait for the graceful shutdown time period, instead that is the max time limit we wait for in worst case. If all the request processing is complete way before, we shutdown fast. This IMO is better for agent performance.

However I do agree that the optimization comes at a cost of code readability. So I am inclined to make the changes to do a simple blind sleep of the configured graceful shutdown time period and then just do shutdown. I will make the changes and update the PR. 

That said, I am keeping this version of the graceful shutdown changes in a git branch if we ever want to implement the optimized waiting logic in future. Link to branch for reference: https://github.com/google/inverting-proxy/commit/7d25d34cc893104092660d0b1625345cc7c443eb